docs: improve scalar index and filtering (#1874)
lancedb/lancedb: docs/src/guides/scalar_index.md

# Building a Scalar Index

Scalar indices organize data by scalar attributes (e.g. numbers, categorical values), enabling fast filtering and sorting of vector data. In vector databases, scalar indices accelerate the retrieval of scalar data associated with vectors, thus enhancing the query performance when searching for vectors that meet certain scalar criteria. 

Similar to many SQL databases, LanceDB supports several types of scalar indices to accelerate search
over scalar columns.

> - `BTREE`: The most common type is BTREE. The index stores a copy of the
>   column in sorted order. A header entry is created for each block of rows
>   (currently the block size is fixed at 4096). These header entries are stored
>   in a separate cacheable structure (a btree). To search for data the header is
>   used to determine which blocks need to be read from disk.

Reviewer: ```suggestion
- `BTREE`: The most common type is BTREE. The index stores a copy of the
  column in sorted order. This sorted copy allows a binary search to be used to
  satisfy queries.
```

Let's avoid some of the details in this overview doc.

          .limit(10)
          .toArray();
        ```
### Update a scalar index
Updating the table data (adding, deleting, or modifying records) requires that you also update the scalar index. This can be done by calling `optimize`, which will trigger an update to the existing scalar index.
=== "Python"

    ```python
    table.add([{"vector": [7, 8], "book_id": 4}])
    table.optimize()
    ```

=== "TypeScript"

    ```typescript
    await tbl.add([{ vector: [7, 8], book_id: 4 }]);
    await tbl.optimize();
    ```

=== "Rust"

    ```rust
    let more_data: Box<dyn RecordBatchReader + Send> = create_some_records()?;
    tbl.add(more_data).execute().await?;
    tbl.optimize(OptimizeAction::All).execute().await?;
    ```

!!! note

>     New data added after creating the scalar index will appear in search results while the index update is still progress, but with increased latency due to a flat search on the unindexed portion. LanceDB Cloud automates this merging process, minimizing the impact on search speed.

Reviewer: ```suggestion
    New data added after creating the scalar index will still appear in search results if optimize is not used, but with increased latency due to a flat search on the unindexed portion. LanceDB Cloud automates the optimize process, minimizing the impact on search speed.
```
